File: opcua-webservice-spring/core/src/main/java/org/opcfoundation/webserver/service/transactions/UaTransactionManager.java
package org.opcfoundation.webserver.service.transactions;

import org.eclipse.milo.opcua.stack.core.UaRuntimeException;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.TreeMap;
import java.util.concurrent.CompletableFuture;

public class UaTransactionManager<T_Request,T_Response> {
    private final List<UaTransaction<T_Request, T_Response>> transactions;
    private int currentTransactionIndex;

    public UaTransactionManager()
    {
        transactions = new ArrayList<>();
        currentTransactionIndex = 0;
    }

    public void addTransaction(UaTransaction<T_Request, T_Response> transaction)
    {
        transactions.add(transaction);
    }

    public List<T_Response> getMergedResults() throws UaRuntimeException
    {
        Map<Integer,T_Response> results = new TreeMap<>();

        for (UaTransaction<T_Request,T_Response> item : transactions)
        {
            results.put(item.getHandleId(), item.getResult());
        }

        return new ArrayList<>(results.values());
    }

    public CompletableFuture<UaTransactionManager<T_Request,T_Response>> execute()
    {
        if (transactions.size() <= currentTransactionIndex || 0 > currentTransactionIndex)
        {
            return CompletableFuture.completedFuture(this);
        }

        return transactions.get(currentTransactionIndex).
                execute().
                thenCompose(V -> executeComplete());
    }

    private CompletableFuture<UaTransactionManager<T_Request,T_Response>> executeComplete()
    {
        currentTransactionIndex++;
        return execute();
    }
}

File: opcua-webservice-node/src/server/service/transactions/base/UaTransactionManager.ts
import { makeUaStatusCode, UaError } from "opcua-webapi-ts";
import { UaTransaction } from "./UaTransaction";
import { StatusCodes } from "opcua-webapi";

export class UaTransactionManager<TRequest, TResponse> {
    private readonly transactions: Array<UaTransaction<TRequest, TResponse>>;
    private currentTransactionIndex: number;

    constructor() {
        this.transactions = [];
        this.currentTransactionIndex = 0;
    }

    addTransaction(transaction: UaTransaction<TRequest, TResponse>): void {
        this.transactions.push(transaction);
    }

    getMergedResults(): Array<TResponse> {
        const results = new Map<number, TResponse>();
        let ret : Array<TResponse> = [];

        for (const item of this.transactions) {
            results.set(item.handleId, item.getResult());
        }

        for (let i = 0; i < results.size; i++) {            
            if (!results.has(i)) throw new UaError(makeUaStatusCode(StatusCodes.BadUnexpectedError));
            ret.push(results.get(i) as TResponse);
        }

        return ret;
    }

    async execute(): Promise<void> {
        if (this.currentTransactionIndex >= this.transactions.length) return;
        await this.transactions[this.currentTransactionIndex].execute();
        this.currentTransactionIndex += 1;
        await this.execute();
    }
}